benchmarks/bench_ml.py

```diff
@@ -0,0 +1,422 @@
+    :param n_calls: a budget of evaluations
+    :param seed_value: random seed, used to set the random number generator in numpy
+    :return: a list of paris (p, f(p)), where p is a dictionary of the form "param name":value,
+            and f(p) is performance measure value achieved by the model for configuration p.
+            Such list contains history of execution of optimization.
+    """
+
+
+    # below seed is necessary for processes which fork at the same time
+    # so that random numbers generated in processes are different
+    np.random.seed(seed_value)
+
+    problem = TesterClass(model, dataset)
+    space = problem.model_class[MODEL_PARAMETERS]
+
+    # initialization
+    estimator = base_estimator(random_state=seed_value)
+    dimensions = [v[0] for k, v in space.items()]
+    solver = Optimizer(dimensions, estimator, random_state=seed_value)
+
+    trace = []
+    best_y = np.inf
+    best_x = None
+
+    # optimization loop
+    for i in range(n_calls):
+        point_list = solver.ask()
+
+        point_dct = {k: v for k, v in zip(space.keys(), point_list)}  # convert list of dimension values to dictionary
+        objective_at_point = -problem.evaluate(point_dct) # the result of "evaluate" is accuracy / r^2, which is the more the better
+
+        if best_y > objective_at_point:
+            best_y = objective_at_point
+            best_x = point_dct
+
+        trace.append((best_x, best_y)) # remember the point, objective pair
+        print("Eval. #"+str(i))
+
+        solver.tell(point_list, objective_at_point)
+
+    return trace
+
+# the function below is necessary as a workaround for the pool.map method requiring map function to take only a single argument
+def _evaluate_optimizer(params):
+    return evaluate_optimizer(*params)
+
+def get_gaussian_surrogate(random_state):
+    return GaussianProcessRegressor(kernel=Matern(), random_state=random_state)
+
+def run(n_calls = 32,
+        n_runs = 1,
+        save_traces = True,
+        parallel_jobs = False):
+    """
+    Main function used to run the experiments.
+
+    :param n_calls: evaluation budget
+    :param n_runs: how many times to repeat the optimization in order to average out noise
+    :param save_traces: whether to save data collected during optimization
+    :param parallel_jobs: whether to run different repeats of optimization in parallel
+    :return: None
+    """
+
+    surrogates = [get_gaussian_surrogate]
+    selected_models = TesterClass.models.keys()
+    selected_datasets = TesterClass.datasets.keys()
+
+    # all the parameter values and objectives collected during execution are stored in list below
+    all_data = {}
+
+    for model in selected_models:
+        all_data[model] = {}
+
+        for dataset in selected_datasets:
+            supports, _ = TesterClass.supports(model, dataset)
+
+            if not supports:
+                continue
+
+            all_data[model][dataset] = {}
+
+            for surrogate in surrogates:
```

Comment:
The reason I did this is so that it is a bit easier later to do visualizations if you want to do comparison of different surrogates on one particular problem, then you just take `all_data[model][dataset]` for some model and dataset, and visualize the results that are stored in there. It might be that some particular dataset / model or subset of those is more representative of some particular optimization task for which you want to choose which surrogate to use, so you might just consider those particular results instead of all results. It might be good also if at some point we decide to split this one large benchmark into smaller ones (just of top of my head)